plenoptic/synthesize/geodesic.py

```diff
@@ -1,155 +1,304 @@
-        self.diff.weight.requires_grad = False
-        self.blur.weight.requires_grad = False
+        self.optimizer.zero_grad()
+        y = self.analyze()
+        repres_path_energy = self.path_energy(y, self.yA, self.yB)
+        loss = repres_path_energy
+        if self.lmbda >= 0:
+            loss = loss + self.lmbda * penalize_range(self.x, (0, 1))
 
-    def analyze(self, x):
+        if loss.item() != loss.item():
+            self.step_energy.pop()
+            raise Exception('found a NaN in the loss during optimization')
 
-        x = torch.cat((self.xA, x, self.xB), 0)
-        x = self.squish(x)
-        self.img = self.mask * x
+        loss.backward()
 
-        return self.model(self.img.unsqueeze(1))
+        # TODO undercomplete case
+        # repres_grad = x.grad
 
-    def objective_function(self, x):
+        # self.optimizer.zero_grad()
+        # signal_path_energy = self.path_energy(self.x, self.xA, self.xB)
+        # signal_grad = x.grad
+        # x.grad = signal_grad - (
+        #           signal_grad @ repres_grad
+        #                         ) / torch.norm(repres_grad) * repres_grad
 
-        z = x.permute(1, 0, 2, 3).unsqueeze(1)
-        loss = 0
-        for s in range(self.n_scales):
-            loss = loss + torch.sum(self.diff(z) ** 2)
-            z = self.blur(z)
+        self.optimizer.step()
+        grad_norm = torch.norm(self.x.grad.data)
+        pbar.set_postfix(OrderedDict([('loss', f'{loss.item():.4e}'),
+                         ('gradient norm', f'{grad_norm:.4e}'),
+                         ('lr', self.optimizer.param_groups[0]['lr'])]))
+        if grad_norm.item() != grad_norm.item():
+            raise Exception('found a NaN in the gradients during optimization')
 
         return loss
 
-    def _optimizer_step(self, i, max_iter):
+    def synthesize(self, max_iter=1000, learning_rate=.001, optimizer='adam',
+                   lmbda=.1, objective='multiscale', seed=0):
+        """ synthesize a geodesic
 
-        self.optimizer.zero_grad()
-        output = self.analyze(self.x)
-        loss = self.objective_function(output)
-        loss.backward()
-        g = self.x.grad.data
-        if math.isnan(loss.item()):
-            raise Exception('found a NaN during optimization')
-        self.optimizer.step()
+        Parameters
+        ----------
+        max_iter: int, optional
+            maximum number of steps taken by the optimization a
 
-        self.t0 = self.t1
-        self.t1 = time.time()
-        pretty_print(i, max_iter, (self.t1 - self.t0), loss.item(), g.norm().item())
+        learning_rate: float, optional
+            controls the step sizes of the search algorithm
 
-        return loss
+        optimizer: str ('adam' or 'sgd') or torch.optim.Optimizer, optional
+            algorithm that will perform the search
 
-    def synthesize(self, max_iter=20, learning_rate=.01, objective='multiscale', seed=0):
+        lmbda: float, optional
```

Comment:
not necessarily in this PR, but I think we should discuss how we want to handle image range across the package: `Metamer` and `MADCompetition` allow for a clamper or assumes it's in the loss function. We should pick some way and be consistent.